packages/core/src/shared-priority-cache/shared-cache.ts

```diff
@@ -0,0 +1,143 @@
+// they expect that the things coming out of the cache are the type they expect (what they put in it)
+// this is not strictly true, as the cache is shared, and other clients may use different types
+// also also - there will not be a 1:1 relation between items and
+type CacheInterface<Key, Value extends Record<string, Resource>> = {
+    get: (k: Key) => Value | undefined
+    has: (k: Key) => boolean
+    unsubscribeFromCache: () => void;
+    setPriorities: (low: Set<Key>, high: Set<Key>) => void;
+}
+type Fetcher = (sig: AbortSignal) => Promise<Resource>
+
+type ClientSpec<Key, Value extends Record<string, Resource>> = {
+    isValue: (v: Record<string, Resource | undefined>) => v is Value
+    cacheKeys: (item: Key) => { [k in keyof Value]: string }
+    onDataArrived?: (cacheKey: string) => void // todo very not helpful
+    fetch: (item: Key) => { [k in keyof Value]: (abort: AbortSignal) => Promise<Resource> }
+}
+type ObjectValue<T extends Record<string, any>> = T extends Record<string, infer Value> ? Value : never;
+
+type KV<T extends Record<string, any>> = readonly [keyof T, ObjectValue<T>]
+
+function entries<T extends Record<string, any>>(t: T): ReadonlyArray<KV<T>> {
+    return Object.entries(t) as ReadonlyArray<KV<T>>
+}
+function mapFields<R extends Record<string, any>, Result>(r: R, fn: (v: ObjectValue<R>) => Result): { [k in keyof R]: Result } {
+    return entries(r).reduce((acc, [k, v]) => ({ ...acc, [k]: fn(v) }), {} as { [k in keyof R]: Result })
+}
+
+type Client = {
+    low: Set<string>
+    high: Set<string>
+    notify: undefined | ((cacheKey: string) => void)
+}
+export class FancySharedCache {
+    private cache: PriorityCache;
+    private clients: Record<string, Client>
+    private importance: Record<string, number>
+    constructor(store: Store<string, Resource>, limitInBytes: number, max_concurrent_fetches: number = 10) {
+        this.importance = {}
+        this.clients = {}
+        this.cache = new PriorityCache(store, (ck) => this.importance[ck] ?? 0, limitInBytes, max_concurrent_fetches, (ck) => this.onCacheEntryArrived(ck))
+    }
+    registerClient<Key, Value extends Record<string, Resource>>(spec: ClientSpec<Key, Value>): CacheInterface<Key, Value> {
+        const id = uniqueId('client')
+        this.clients[id] = { low: new Set(), high: new Set(), notify: spec.onDataArrived }
+
+        const makeCacheEntries = (item: Key) => {
+            const keys = spec.cacheKeys(item) as Record<string, string>
+            const fetchers = spec.fetch(item)
+            return Object.keys(keys).map((sk) => ({ cacheKey: keys[sk], fetcher: fetchers[sk] }))
+        }
+
+        const priorityDelta = (key: string, pri: 0 | 1 | 2, low: Set<string>, high: Set<string>) => {
+            return pri - (high.has(key) ? 2 : (low.has(key) ? 1 : 0))
+        }
+        const setPriorities = (low: Set<Key>, high: Set<Key>) => {
+            const newLow = new Set<string>()
```

Comment:
here is the meat of the new approach - every time a client updates its priorities, the system has to re-balance the heap of all priorities. This also has a bug due to how Set() deals with equality, and thus membership

todo: rewrite and test more

Reply:
I've re-written this - now I'm going to try to write some tests for it